## Import some necessary lib

In [4]:
import pandas as pd

In [5]:
path = ("C:/Users/ASUS/Documents/Project/Momo case study/Data/raw/momo_top_up.xlsx")

### Reading file from excel sheets

In [6]:
df_transactions = pd.read_excel(path, sheet_name=0)
df_user_info = pd.read_excel(path, sheet_name=2)
df_commision = pd.read_excel(path, sheet_name= 1)

# Transactions

In [7]:
print("Size of df Transactions")
print(df_transactions.shape)

Size of df Transactions
(13495, 6)


## 1. Transactions
Momo top-up transaction table in 2020
- user_id: user ID
- order_id: Transaction id (unique)
- Date: Transaction date 
- Amount: Top-up amount  (VNĐ)
- Merchant_id: Telecom ID (12=Viettel, 13=Vinaphone,...)
- Purchase_status: Pursche status (Only available for some transaction)

In [ ]:
print("Information about the transactions dataframe")
df_transactions.info()
print("-----------------------")

Information about the `transactions` dataframe
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13495 entries, 0 to 13494
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   user_id          13495 non-null  int64 
 1   order_id         13495 non-null  int64 
 2   Date             13495 non-null  object
 3   Amount           13495 non-null  object
 4   Merchant_id      13495 non-null  int64 
 5   Purchase_status  2235 non-null   object
dtypes: int64(3), object(3)
memory usage: 632.7+ KB
-----------------------


### Dataframe values preview

In [9]:
df_transactions.head(3)

,user_id,order_id,Date,Amount,Merchant_id,Purchase_status
0,21269588,4169517626,2020-01-01,"10,000",13,NaN
1,28097592,4170276686,2020-01-01,"20,000",13,NaN
2,47435144,4166729310,2020-01-01,"10,000",12,NaN


### Checking missing values

In [16]:
missing = df_transactions.isnull().sum()
missing_pct = df_transactions.isnull().mean() * 100
pd.DataFrame({'Missing Count': missing, 'Missing %': missing_pct.round(2)})

,Missing Count,Missing %
user_id,0,0.00
order_id,0,0.00
Date,0,0.00
Amount,0,0.00
Merchant_id,0,0.00
Purchase_status,11260,83.44


### Checking duplicates values

In [11]:
print("Duplicates rows:", df_transactions.duplicated().sum())
print("Order_id duplicates", df_transactions['order_id'].duplicated().sum())


Duplicates rows: 0
Order_id duplicates 0


### Unique values (categorical columns)

In [13]:
# Xem có bao nhiêu giá trị unique
for col in ['Merchant_id', 'Purchase_status']:
    print(f"\n{col} ({df_transactions[col].nunique()} unique values):")
    print(df_transactions[col].value_counts(dropna=False))



Merchant_id (5 unique values):
Merchant_id
12    6777
13    3783
14    2321
15     611
16       3
Name: count, dtype: int64

Purchase_status (1 unique values):
Purchase_status
NaN       11260
Mua hộ     2235
Name: count, dtype: int64


### Descriptive statistics

In [14]:
df_transactions.describe(include='all')

,user_id,order_id,Date,Amount,Merchant_id,Purchase_status
count,1.349500e+04,1.349500e+04,13495,13495,13495.000000,2235
unique,NaN,NaN,366,16,NaN,1
top,NaN,NaN,2020-06-30,"20,000",NaN,Mua hộ
freq,NaN,NaN,79,3357,NaN,2235
mean,4.021140e+07,6.385662e+09,NaN,NaN,12.761023,NaN
std,1.368218e+07,1.289348e+09,NaN,NaN,0.894556,NaN
min,1.138060e+05,4.166563e+09,NaN,NaN,12.000000,NaN
25%,3.597480e+07,5.303238e+09,NaN,NaN,12.000000,NaN
50%,4.346840e+07,6.274443e+09,NaN,NaN,12.000000,NaN
75%,4.940701e+07,7.502845e+09,NaN,NaN,13.000000,NaN


### Date range

In [15]:
# Because date is an object type so i use min,max function
print("From:", df_transactions['Date'].min())
print("To:", df_transactions['Date'].max())

From: 2020-01-01
To: 30/9/2020


## Data Quality Issues
| Column | Issue | Action |
|---|---|---|
| `Amount` | String format with comma | Convert to int |
| `Date` | Stored as object ,  | Parse to datetime |
| `Purchase_status` | 83% missing | Investigate in cleaning |

# User_info

In [17]:
print("Size of df User_info")
print(df_user_info.shape)

Size of df User_info
(13428, 5)


## 2. User Info
User Info using Top-up service of Momo in 2020 
- User_id : ID of user
- First_trans_date : First transaction date
- Location : Location of user
- Age : Age of user
- Gender : Gender of user


In [18]:
print("Information about the user_info dataframe")
df_user_info.info()
print("-----------------------")

Information about the user_info dataframe
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13428 entries, 0 to 13427
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   User_id          13428 non-null  int64 
 1   First_tran_date  13428 non-null  object
 2   Location         13428 non-null  object
 3   Age              13428 non-null  object
 4   Gender           13428 non-null  object
dtypes: int64(1), object(4)
memory usage: 524.7+ KB
-----------------------


### Dataframe preview

In [30]:
df_user_info.head(3)

,User_id,First_tran_date,Location,Age,Gender
0,41654498,2018-05-02,HCMC,33_to_37,MALE
1,51276281,2019-12-08,Other Cities,unknown,MALE
2,49152375,2019-12-20,Other Cities,23_to_27,MALE


### Checking missing values

In [19]:
missing_userinfo = df_user_info.isnull().sum()
missing_userinfo_pct = df_user_info.isnull().mean() *  100
pd.DataFrame({'Missing Count': missing_userinfo, 'Missing %': missing_userinfo_pct.round(2)})

,Missing Count,Missing %
User_id,0,0.0
First_tran_date,0,0.0
Location,0,0.0
Age,0,0.0
Gender,0,0.0


### Checking duplicates values


In [21]:
print("Rows duplicate:",df_user_info.duplicated().sum())
print("User_id duplicate:",df_user_info["User_id"].duplicated().sum())

Rows duplicate: 0
User_id duplicate: 38


### Unique values (categorical columns)


In [25]:
for col in ['Location', 'Age', 'Gender']:
    print(f"\n{col}({df_user_info[col].nunique()})")
    print(df_user_info[col].value_counts(dropna=False))


Location(6)
Location
Other Cities        6022
HCMC                4112
HN                  1437
Other               1033
Unknown              761
Ho Chi Minh City      63
Name: count, dtype: int64

Age(6)
Age
23_to_27    3421
28_to_32    2796
unknown     2521
33_to_37    1745
18_to_22    1695
>37         1250
Name: count, dtype: int64

Gender(8)
Gender
MALE      6282
FEMALE    3407
Nam       1370
Nữ        1248
female     903
M           92
male        71
f           55
Name: count, dtype: int64


### Descriptive statistics


In [26]:
df_user_info.describe(include='all')

,User_id,First_tran_date,Location,Age,Gender
count,1.342800e+04,13428,13428,13428,13428
unique,NaN,1590,6,6,8
top,NaN,2020-02-14,Other Cities,23_to_27,MALE
freq,NaN,45,6022,3421,6282
mean,4.022728e+07,NaN,NaN,NaN,NaN
std,1.367326e+07,NaN,NaN,NaN,NaN
min,1.138060e+05,NaN,NaN,NaN,NaN
25%,3.602175e+07,NaN,NaN,NaN,NaN
50%,4.347301e+07,NaN,NaN,NaN,NaN
75%,4.941236e+07,NaN,NaN,NaN,NaN


### Date range

In [29]:
#Because the First_tran_date isn't datetime, so i use min , max function
print("From:",df_user_info['First_tran_date'].min())
print("To:",df_user_info['First_tran_date'].max())

From: 2016-01-01
To: 9920-12-16


## Data Quality Issues - User Info

| Column | Issue | Proposed Action |
|---|---|---|
| `User_id` | 38 duplicated rows | Remove duplicates keeping the latest/valid record |
| `First_tran_date` | Out-of-range value (e.g., year 9920) | Replace anomalous dates with median/mean or drop |
| `Location` | Inconsistent naming (`Ho Chi Minh City` vs `HCMC`) | Standardize values (e.g., map to `HCMC`) |
| `Age` | Contains 2,521 `unknown` values | Treat as missing values (impute or keep as a separate segment) |
| `Gender` | Inconsistent format (`MALE`, `male`, `Nam`, `Nữ`, etc.) | Standardize to uppercase English (`MALE`, `FEMALE`) |


## 3. Commission (Lookup Table)
This is a supporting table (dimension table) that stores the commission rate for each telecom provider. It will be joined with the `Transactions` table via `Merchant_id` to calculate the final revenue.
| Column | Type | Description |
|---|---|---|
| Merchant_id | int | Telecom provider ID (Primary Key used for joining) |
| Rate_pct | float | Commission rate percentage charged by MoMo

In [ ]:
print(df_commision.head())

  Merchant_name  Merchant_id  Rate_pct
0       Viettel           12         2
1      Mobifone           13         3
2     Vinaphone           14         4
3  Vietnamobile           15         4
4       Gmobile           16         4


# Data Understanding Summary & Next Steps
After exploring the dataset, here is the summary of the data quality and the required actions for the next processing stage (**Data Cleaning**):
## 1. Transactions Table
*   **Datatype Conversion**: `Amount` is a string with commas -> Needs conversion to numeric (int).
*   **Datatype Conversion**: `Date` is stored as an object -> Needs conversion to `datetime`.
*   **Missing Values**: `Purchase_status` has 83.4% missing values -> Need a strategy to handle them (e.g., fill with "Normal" or categorize).
## 2. User Info Table
*   **Duplicates**: 38 duplicated `User_id`s found -> Needs deduplication (keeping the valid/latest record).
*   **Anomalous Dates**: `First_tran_date` contains out-of-range years (e.g., 9920) and is stored as an object -> Investigate, fix outliers, and convert to `datetime`.
*   **Inconsistent Data (Gender)**: Gender column is messy (`MALE`, `male`, `Nữ`, `Nam`, etc.) -> Standardize formatting (e.g., `Male`/`Female`).
*   **Inconsistent Data (Location)**: Overlapping categories (`Ho Chi Minh City` vs `HCMC`) -> Map to standardized names.
*   **Implicit Missing Values**: `Age` contains 2,521 `unknown` values -> Decide whether to treat as a separate cohort or impute.
## 3. Commission Table
*   **Datatype Check**: Verify if `Rate_pct` is stored as percentage percentages (e.g., `0.05` instead of `5`) to ensure accurate revenue calculation later. 
*   **Quality**: No duplicate keys found. Clean and ready to be used as a lookup table.
### Move to `data_cleaning.ipynb` to execute the above actions.